src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs

```diff
@@ -637,57 +639,41 @@ private unsafe string[] GetValueNamesCore(int values)
                                 throw new IOException(SR.Arg_RegGetOverflowBug, e);
                             }
                         }
-                        char[] blob = new char[datasize / 2];
 
-                        Interop.Advapi32.RegQueryValueEx(_hkey, name, null, ref type, blob, ref datasize);
-                        if (blob.Length > 0 && blob[blob.Length - 1] == (char)0)
-                        {
-                            data = new string(blob, 0, blob.Length - 1);
-                        }
-                        else
+                        char[]? pooledArray = null;
+                        int blobLength = datasize / 2;
+                        Span<char> blob = blobLength <= 128 ? // arbitrary limit
+                            stackalloc char[blobLength] :
+                            (pooledArray = ArrayPool<char>.Shared.Rent(blobLength));
+                        blob.Clear();
+
+                        fixed (char* blobPtr = &MemoryMarshal.GetReference(blob))
                         {
-                            // in the very unlikely case the data is missing null termination,
-                            // pass in the whole char[] to prevent truncating a character
-                            data = new string(blob);
+                            Interop.Advapi32.RegQueryValueEx(_hkey, name, null, ref type, blobPtr, ref datasize);
                         }
-                    }
-                    break;
 
-                case Interop.Advapi32.RegistryValues.REG_EXPAND_SZ:
-                    {
-                        if (datasize % 2 == 1)
+                        if (!blob.IsEmpty && blob[blobLength - 1] == '\0')
```

Comment:
This routine seems to take it for granted that datasize in the first call was accurate.  Maybe registry does guarantee that, but not all Win32 does (at least some of Win32 Crypto reports "here's a number that's guaranteed big enough, but the answer might be smaller when you make me do actual work").  Seems like this would be more robust if it recomputed based on `datasize` after the call.

But maybe it's fine.  After all, we've had this pattern for 20 years...